### PR TITLE
Hough_Line_Probabilistic Added

### PR DESCRIPTION
**Pull request by vernwalrahul**
Implemented Hough_Line_Probabilistic in hough_transform.jl as proposed in the following paper:
[Link](http://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=786993).

Following are some of the results:
1canny
1hough_p
2canny
2hough_p

### CODE REVIEW

**Comment by zygmuntszpak** [comment]
Thank you for taking the time to make a contribution. I've downloaded the paper so I can follow the code better.

**Comment by vernwalrahul** [comment]
I have updated the code adding comments and reference. Can you please review this PR?
ping @zygmuntszpak  @timholy

**Comment by zygmuntszpak** [comment]
Could you also add some tests for this function please? Thanks for adding the comments---they will make it easier to verify the implementation. I haven't had the time yet to think about this more deeply, but at first glance your function seems very long and hence difficult to parse visually.  Perhaps some of the steps ought to go into  separate "helper" functions?

**Comment by vernwalrahul** [comment]
@zygmuntszpak  
Thanks for the comments. I will try adding separate helper functions.
And since the algorithm is dependent on randomly selected points, it may turn up with different line segments each time we run it. I wonder if adding test cases will suffice the purpose.

**Comment by zygmuntszpak** [comment]
For the most basic validation you could organise your test so that there is only one line segment and maybe a few other scattered random points. You can seed the random number generator with a particular seed so that every time the experiment is run the same set of random points are generated. Presumably the algorithm should always find the only line segment in the example? 

You could then try another test with two line segments. 

Also, instead  of `error(...)` use `throw(ArgumentError(...))`

**Comment by vernwalrahul** [comment]
@zygmuntszpak  I have added the required helper functions along with few test cases, as suggested.

**Comment by zygmuntszpak** [comment]
Thank you Rahul. Could you please move the functions that are currently defined inside `hough_line_probabilistic` outside of `hough_line_probabilistic`. For example,  move `function collect_points()` outside of `hough_line_probabilistic`. The aim is to make  `hough_line_probabilistic` as short as possible so that it can, ideally,  fit onto a single page. The brevity will help people who are encountering the code for the first time parse the logic of the algorithm.

**Comment by vernwalrahul** [comment]
@zygmuntszpak Sure, I can do that. But that will increase the number of arguments, I think. How about adding a separate file as "hough_line_submodules.jl" ? I can import required data and functions from there only. There is one more function in hough_line_standard written in nested form.

**Comment by zygmuntszpak** [comment]
I'm not sure I fully understand your separate file idea. I don't know to how best to resolve this, but I do feel that the code is one very long piece which gets in the way of interpreting its intent. Perhaps you could group some of the parameters in a Type if you feel the argument list to the helper functions is getting too long?

**Comment by vernwalrahul** [comment]
@zygmuntszpak  I have moved the nested functions outside, passing the arguments after wrapping the various parameters.

**Comment by zygmuntszpak** [comment]
I've checked your branch out and used the ["Traceur.jl"](https://github.com/MikeInnes/Traceur.jl) package to check for Type stability. In particular, I ran `@trace lines = hough_line_probabilistic(img, 1, linspace(0,π,180),9,7,15,4)` using the `img` associated with the `#for a square image` part of your test code. 

The `@trace` command found some issues. Here is the output of the analysis:
```
@trace lines = hough_line_probabilistic(img, 1, linspace(0,π,180),9,7,15,4)
(ImageFeatures.update_accumulator)(::ImageFeatures.Param, ::Tuple{Int64,Int64}, ::Array{Float64,1}, ::Array{Float64,1}) at /home/zygmunt/.julia/v0.6/ImageFeatures/src/houghtransform.jl:188
  dist is assigned as Int64 at line 191
  dist is assigned as Float64 at line 192
  dist is assigned as Int64 at line 193
  val is assigned as Any at line 195
  max_val is assigned as Int64 at line 189
  max_val is assigned as Any at line 197
  dynamic dispatch to (ImageFeatures.getindex)((Core.getfield)(params, :accumulator_matrix), (Base.add_int)(n, 1), (Base.add_int)(_7::Int64, 1)) at line 194
  dynamic dispatch to (ImageFeatures.getindex)((Core.getfield)(params, :accumulator_matrix), (Base.add_int)(n, 1), (Base.add_int)(_7::Int64, 1)) + 1 at line 194
  dynamic dispatch to (ImageFeatures.getindex)((Core.getfield)(params, :accumulator_matrix), (Base.add_int)(n, 1), (Base.add_int)(_7::Int64, 1)) at line 195
  dynamic dispatch to max_val < val at line 196
  returns Tuple{Int64,Any}
(Base.indexed_next)(::Tuple{Int64,Bool}, ::Int64, ::Int64) at tuple.jl:54
  returns Tuple{Union{Bool, Int64},Int64}
(ImageFeatures.pass_1)(::ImageFeatures.Param, ::ImageFeatures.Sample, ::Int64) at /home/zygmunt/.julia/v0.6/ImageFeatures/src/houghtransform.jl:206
  h is assigned as Any at line 207
  w is assigned as Any at line 207
  #temp# is assigned as Any at line 207
  #temp# is assigned as Any at line 207
  #temp# is assigned as Any at line 207
  #temp# is assigned as Bool at line 232
  #temp# is assigned as Any at line 232
  #temp# is assigned as Bool at line 232
  #temp# is assigned as Any at line 232
  #temp# is assigned as Any at line 232
  #temp# is assigned as Any at line 232
  dynamic dispatch to (ImageFeatures.size)((Core.getfield)(params, :mask)) at line 207
  dynamic dispatch to (Base.start)((ImageFeatures.size)((Core.getfield)(params, :mask))) at line 207
  dynamic dispatch to (Base.indexed_next)((ImageFeatures.size)((Core.getfield)(params, :mask)), 1, #temp#) at line 207
  dynamic dispatch to (Base.indexed_next)((ImageFeatures.size)((Core.getfield)(params, :mask)), 2, #temp#) at line 207
  dynamic dispatch to j1 >= w at line 232
  dynamic dispatch to i1 >= h at line 232
  dynamic dispatch to (ImageFeatures.getindex)((Core.getfield)(params, :mask), (Base.add_int)(i1, 1), (Base.add_int)(j1, 1)) at line 238
(ImageFeatures.hough_line_probabilistic)(::Array{Bool,2}, ::Int64, ::StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}, ::Int64, ::Int64, ::Int64, ::Int64) at /home/zygmunt/.julia/v0.6/ImageFeatures/src/houghtransform.jl:304
  max_val is assigned as Any at line 345
  dynamic dispatch to (ImageFeatures.getindex)((Core.getfield)(params, :mask), (Base.getfield)(point, 1), (Base.getfield)(point, 2)) at line 340
  dynamic dispatch to !((ImageFeatures.getindex)((Core.getfield)(params, :mask), (Base.getfield)(point, 1), (Base.getfield)(point, 2))) at line 340
  dynamic dispatch to max_val < (Core.getfield)(params, :threshold) at line 348
4-element Array{NTuple{4,Int64},1}:
 (3, 3, 3, 14)
 (14, 3, 4, 3)
 (14, 14, 4, 14)
 (14, 4, 14, 13)

```
Note, for example, the lines: 
```
  dist is assigned as Int64 at line 191
  dist is assigned as Float64 at line 192
  dist is assigned as Int64 at line 193
```
The problem is that `dist` is being converted to a `Float64` on line 192 by virtue of the fact that `params.constadd` is declared as a `Float64`. 

In a nutshell, you don't want to to be changing the Type of a variable and furthermore, you don't want a variable to be identified as `Any`.  See the discussion on the [performance tips](https://docs.julialang.org/en/stable/manual/performance-tips/) in the Julia documentation. I'm not entirely sure what the cause of `#temp#` type instability is, but perhaps you could sort out the non-temporary variables in the meantime. 

Thank you for your valuable time and effort.

**Comment by vernwalrahul** [comment]
@zygmuntszpak  Thanks for the info. I didn't know about this trace command.
I have corrected the code accordingly, but am unable to check for the issues. I am not sure if I have cloned the right package. Is it [https://github.com/MikeInnes/Traceur.jl](https://github.com/MikeInnes/Traceur.jl)?

This one is showing the following error:
screenshot from 2018-03-22 17-59-50

**Comment by zygmuntszpak** [comment]
Hmm..not sure what is going on there. Did you do `Pkg.add("Traceur")` and then 
```
using Traceur
# TODO: define img

@trace lines = hough_line_probabilistic(img, 1, linspace(0,π,180),9,7,15,4) 
```

**Comment by vernwalrahul** [comment]
@zygmuntszpak Yes, I did exactly the same thing. It's Traceur v0.1.1 here. I don't know why it's showing the same error with hough_transform_standard as well.
screenshot from 2018-03-24 13-33-27

**Comment by zygmuntszpak** [comment]
That's odd, I am using the same version of Traceur and it works for me. The latest output is:

```
@trace hough_line_probabilistic(img, 1, linspace(0,π,180),9,7,15,4)
(ImageFeatures.update_accumulator)(::ImageFeatures.Param, ::Tuple{Int64,Int64}, ::Array{Float64,1}, ::Array{Float64,1}) at /home/zygmunt/.julia/v0.6/ImageFeatures/src/houghtransform.jl:188
  dist is assigned as Float64 at line 191
  dist is assigned as Float64 at line 192
  dist is assigned as Int64 at line 193
  val is assigned as Any at line 195
  max_val is assigned as Int64 at line 189
  max_val is assigned as Any at line 197
  dynamic dispatch to (ImageFeatures.getindex)((Core.getfield)(params, :accumulator_matrix), (Base.add_int)(n, 1), (Base.add_int)(_7::Int64, 1)) at line 194
  dynamic dispatch to (ImageFeatures.getindex)((Core.getfield)(params, :accumulator_matrix), (Base.add_int)(n, 1), (Base.add_int)(_7::Int64, 1)) + 1 at line 194
  dynamic dispatch to (ImageFeatures.getindex)((Core.getfield)(params, :accumulator_matrix), (Base.add_int)(n, 1), (Base.add_int)(_7::Int64, 1)) at line 195
  dynamic dispatch to max_val < val at line 196
  returns Tuple{Int64,Any}
(Base.indexed_next)(::Tuple{Int64,Bool}, ::Int64, ::Int64) at tuple.jl:54
  returns Tuple{Union{Bool, Int64},Int64}
(ImageFeatures.pass_1)(::Array{Bool,2}, ::ImageFeatures.Param, ::ImageFeatures.Sample, ::Int64) at /home/zygmunt/.julia/v0.6/ImageFeatures/src/houghtransform.jl:206
  dynamic dispatch to (ImageFeatures.getindex)((Core.getfield)(params, :mask), (Base.add_int)(i1, 1), (Base.add_int)(j1, 1)) at line 238
(ImageFeatures.hough_line_probabilistic)(::Array{Bool,2}, ::Int64, ::StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}, ::Int64, ::Int64, ::Int64, ::Int64) at /home/zygmunt/.julia/v0.6/ImageFeatures/src/houghtransform.jl:304
  max_val is assigned as Any at line 345
  dynamic dispatch to (ImageFeatures.getindex)((Core.getfield)(params, :mask), (Base.getfield)(point, 1), (Base.getfield)(point, 2)) at line 340
  dynamic dispatch to !((ImageFeatures.getindex)((Core.getfield)(params, :mask), (Base.getfield)(point, 1), (Base.getfield)(point, 2))) at line 340
  dynamic dispatch to max_val < (Core.getfield)(params, :threshold) at line 348
4-element Array{NTuple{4,Int64},1}:
 (14, 3, 3, 3)
 (3, 4, 3, 14)
 (14, 14, 4, 14)
 (14, 4, 14, 12)
```

You can also try `@code_warntype hough_line_probabilistic(img, 1, linspace(0,π,180),9,7,15,4)`

But the output of that in your case is very long. It does pick up that `maxval` and `val` is any. We might have to call in the Calvary on this one ;) @kmsquire When you get a chance, could you please review this code? We're not sure how to resolve the type stability issues.
